### PR TITLE
net: change enc28j60 to new SPI API

### PR DESCRIPTION
**Pull request by locomuco**
Signed-off-by: Matthias Boesl <matthias.boesl@gmail.com>

### CODE REVIEW

File: drivers/ethernet/eth_enc28j60.c [hunk 1]
```diff
@@ -43,12 +50,19 @@ static void eth_enc28j60_set_bank(struct device *dev, u16_t reg_addr)
 	tx_buf[0] = ENC28J60_SPI_RCR | ENC28J60_REG_ECON1;
 	tx_buf[1] = 0x0;
 
-	spi_transceive(context->spi, tx_buf, 2, tx_buf, 2);
+	struct spi_buf tx_bufs[] = {
```
Comment (tbursztyka) [inline drivers/ethernet/eth_enc28j60.c]: Always declare variables at the beginning of a code block, not in the middle, so move this up.
Apply that rule everywhere.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Apply also the necessary changes pin-pointed by checkpatch (80chars limit etc...)

**Comment by locomuco** [comment]
@tbursztyka the PR is updated to support legacy and new SPI interface now

File: drivers/spi/Kconfig [hunk 2]
```diff
@@ -25,7 +25,7 @@ menuconfig SPI
 
 config SPI_QMSI
 	bool "QMSI driver for SPI controller"
-	depends on SPI && QMSI
+	depends on SPI && QMSI && SPI_LEGACY_API
```
Comment (tbursztyka) [inline drivers/spi/Kconfig]: These Kconfig fixes should be in another patch

File: drivers/ethernet/eth_enc28j60_priv.h [hunk 3]
```diff
@@ -229,7 +231,7 @@ struct eth_enc28j60_runtime {
 			      CONFIG_ETH_ENC28J60_RX_THREAD_STACK_SIZE);
 	struct k_thread thread;
 	struct device *gpio;
-	struct device *spi;
+	struct spi_config spi_cfg;
 	struct gpio_callback gpio_cb;
```
Comment (tbursztyka) [inline drivers/ethernet/eth_enc28j60_priv.h]: You will need #ifdef/#else/#endif instead of removing *spi

File: drivers/ethernet/eth_enc28j60.c [hunk 4]
```diff
@@ -360,6 +491,14 @@ static int eth_enc28j60_init(struct device *dev)
...
 	}
 
+	/* Initialize GPIO */
+	if (gpio_pin_configure(enc_spi_cs.gpio_dev, config->spi_cs_pin,
+			       (GPIO_DIR_OUT | GPIO_PUD_PULL_DOWN))) {
+		SYS_LOG_ERR("Unable to configure GPIO pin %u",
+			    config->gpio_pin);
+		return -EINVAL;
+	}
+
```
Comment (tbursztyka) [inline drivers/ethernet/eth_enc28j60.c]: You don't need to configure the gpio cs by yourself, neither in spi legacy nor the new one.

File: drivers/ethernet/eth_enc28j60.c [hunk 5]
```diff
@@ -335,19 +443,42 @@ static int eth_enc28j60_init(struct device *dev)
...
-
+#else
+	/* SPI config */
+	context->spi_cfg.operation = SPI_WORD_SET(8);
+	context->spi_cfg.frequency = config->spi_freq;
+	context->spi_cfg.slave = config->spi_slave;
+	context->spi_cfg.vendor = 1;
+	context->spi_cfg.cs = &enc_spi_cs;
+	enc_spi_cs.gpio_pin = config->spi_cs_pin;
+
```
Comment (tbursztyka) [inline drivers/ethernet/eth_enc28j60.c]: this will need no handle the case when config->spi_cs_port is just "" (aka: undefined), which means there is no need of enc_spi_cs

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: drivers/spi/Kconfig (hunk 2 above)
Comment (locomuco) [reply drivers/spi/Kconfig]: what's platforms are exactly tested in the shippable builds for the enc28j60?

File: drivers/ethernet/eth_enc28j60.c [hunk 6]
```diff
@@ -360,6 +493,16 @@ static int eth_enc28j60_init(struct device *dev)
 		return -EINVAL;
 	}
 
+#ifndef CONFIG_SPI_LEGACY_API
```
Comment (locomuco) [inline drivers/ethernet/eth_enc28j60.c]: @tbursztyka  ic checked that, and didn't work without this statement on HW, should this pin be configured in the SPI layer?

File: drivers/ethernet/eth_enc28j60.c (hunk 5 above)
Comment (locomuco) [reply drivers/ethernet/eth_enc28j60.c]: will do

File: drivers/ethernet/eth_enc28j60.c [hunk 7]
```diff
@@ -360,6 +493,16 @@ static int eth_enc28j60_init(struct device *dev)
...
+	/* Initialize GPIO */
+	if (gpio_pin_configure(enc_spi_cs.gpio_dev, config->spi_cs_pin,
+			       (GPIO_DIR_OUT | GPIO_PUD_PULL_DOWN))) {
+		SYS_LOG_ERR("Unable to configure GPIO pin %u",
+			    config->gpio_pin);
+		return -EINVAL;
+	}
+#endif
+
```
Comment (tbursztyka) [inline drivers/ethernet/eth_enc28j60.c]: [HOW COME github let me delete your comment, by mistake? how smart is this... sorry ]

spi_context_cs_configure() should do it, on spi driver side.

Could it be your board needs something specific? I see this GPIO_PUD_PULL_DOWN, looks like it's specific. (it's not supported by all gpio controller)

I could make a quick patch so extra gpio directives can be passed to gpio_pin_configure() in drivers/spi/spi_context.h

Comment (locomuco) [reply drivers/ethernet/eth_enc28j60.c]: no problem :-) , i might move that part into my board folder, thanks

**Comment by locomuco** [comment]
@tbursztyka now everything should be addressed

**Review by tbursztyka — DISMISSED** [review]
Let's call it a go.

When legacy API will be finally dropped, it would be interesting to change the logic of writing/reading a packet in enc28j60's memory: trying to make it in one go, instead of looping currently by MAX_BUFFER_LENGTH (in "segments"). New spi api is exactly meant for this: avoiding multiple calls and memory copies.
Doing it right now would entangle way more #ifdefs everywhere so let's postpone this.

File: drivers/ethernet/eth_enc28j60.c [hunk 8]
```diff
@@ -23,88 +23,153 @@
 
 #define D10D24S 11
 
+#ifndef CONFIG_SPI_LEGACY_API
+static struct spi_cs_control enc_spi_cs = {
+	.gpio_pin = 0,
+	.delay = 0
+};
+#endif
```
Comment (nashif) [inline drivers/ethernet/eth_enc28j60.c]: if we are moving this driver to new API, why are we checking for legacy?

**Review by nashif — CHANGES_REQUESTED** [review]

File: drivers/ethernet/eth_enc28j60.c (hunk 8 above)
Comment (tbursztyka) [reply drivers/ethernet/eth_enc28j60.c]: The patch makes bot SPI API supported at built time.

Comment (locomuco) [reply drivers/ethernet/eth_enc28j60.c]: @nashif the driver tests are still using platforms not supporting the new SPI interface, at the moment supporting both API's gives us a higher test coverage

Comment (nashif) [reply drivers/ethernet/eth_enc28j60.c]: > The patch makes both SPI API supported, choice being made at built time.

we should only support one API, and if a platform still support the old API, it will never be moved if we keep this running in this mode and we will end up with 2 APIs for a long time.

I prefer to have this support 1 single API and accelerate the move to the new API with a deadline.

Comment (locomuco) [reply drivers/ethernet/eth_enc28j60.c]: @nashif no problem, I agree with that, but there is still the open question, how to handle the tests then?

**Comment by locomuco** [comment]
superseeded by #5921
